File: ch02/nodejs/2.3.callstack.js
const callstack = [];
let returnValue;

callstack.push({
    'returnAddr': 'start',
    'number': 5
});


while(callstack.length > 0) {

    const { returnAddr, number } = callstack[callstack.length - 1]
    console.log(callstack)
    if (returnAddr === 'start') {
        if (number === 1) {
            returnValue = 1;
            callstack.pop();
            continue;
        } else {
            callstack[callstack.length - 1].returnAddr = 'after recursive call';
            callstack.push({
                'returnAddr': 'start',
                'number': number - 1
            })

            continue
        }
    }

    if (returnAddr === 'after recursive call') {
        returnValue *= number;

        callstack.pop();
        console.log(returnValue)
        continue
    }
}

console.log(returnValue)

File: ch02/python/2.3.callstack.py
callStack = []
callStack.append({'returnAddr': 'start','number':5})
returnValue = None

while len(callStack) > 0:

    number = callStack[-1]['number']
    returnAddr = callStack[-1]['returnAddr']

    if returnAddr == 'start':
        if number == 1:
         # Basic Situation
            returnValue = 1
            callStack.pop() # "Return" from "function call".

            continue
        else:
            # Recursive
            callStack[-1]['returnAddr'] = 'after recursive call'

            # "Call" the "factorial() function":
            callStack.append({'returnAddr': 'start', 'number': number - 1})
            continue
    elif returnAddr == 'after recursive call':
       returnValue = number * returnValue
       callStack.pop()
       continue

print(returnValue)